### PR TITLE
8336479: Provide Process.waitFor(Duration)

### PR DESCRIPTION
**Pull request by naotoj**
Proposing a new overload method for `Process#waitFor()` which takes a `Duration` for the timeout value. This will reduce the possibility for making mistakes with the `TimeUnit` in the other overload method. A corresponding CSR has also been drafted.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8336494](https://bugs.openjdk.org/browse/JDK-8336494) to be approved

### Issues
 * [JDK-8336479](https://bugs.openjdk.org/browse/JDK-8336479): Provide Process.waitFor(Duration) (**Enhancement** - P4)
 * [JDK-8336494](https://bugs.openjdk.org/browse/JDK-8336494): Provide Process.waitFor(Duration) (**CSR**)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20220/head:pull/20220` \
`$ git checkout pull/20220`

Update a local copy of the PR: \
`$ git checkout pull/20220` \
`$ git pull https://git.openjdk.org/jdk.git pull/20220/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20220`

View PR using the GUI difftool: \
`$ git pr show -t 20220`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20220.diff">https://git.openjdk.org/jdk/pull/20220.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20220#issuecomment-2233936317)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/Process.java [hunk 1]
```diff
@@ -456,14 +457,54 @@ public final BufferedWriter outputWriter(Charset charset) {
...
+    public boolean waitFor(Duration duration)
+            throws InterruptedException {
+        Objects.requireNonNull(duration, "duration"); // throw NPE before other conditions
+
+        if (hasExited())
+            return true;
+        if (duration.isZero() || duration.isNegative())
+            return false;
+
+        return waitForNanos(TimeUnit.NANOSECONDS.convert(duration));
```
Comment (liach) [inline src/java.base/share/classes/java/lang/Process.java:504]: `waitFor` can be overridden by pre-24 subclasses to provide a better implementation while `waitForNanos` couldn't. Is it reasonable for our default implementation to delegate to `waitFor(long, TimeUnit)` that existing user classes have better implementations for?

File: src/java.base/share/classes/java/lang/Process.java [hunk 2]
```diff
@@ -456,14 +457,54 @@ public final BufferedWriter outputWriter(Charset charset) {
...
+     * Causes the current thread to wait, if necessary, until the
+     * process represented by this {@code Process} object has
+     * terminated, or the specified waiting duration elapses.
+     *
+     * <p>If the process has already terminated then this method returns
+     * immediately with the value {@code true}.  If the process has not
+     * terminated and the duration is not positive, then
+     * this method returns immediately with the value {@code false}.
+     *
+     * <p>The default implementation of this method polls the {@code exitValue}
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/Process.java:481]: There was discussion in the CSR on moving this to an `@implSpec`, are you going to do that change?

File: src/java.base/share/classes/java/lang/Process.java (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/Process.java:504]: > waitFor can be overridden by pre-24 subclasses to provide a better implementation while ...

It doesn't really make sense to extend Process, except maybe for mocking or other testing. Process is really for JDK implementations, it's just historical the constructor is public. Just saying that the compatibility concerns with adding methods aren't significant here.

Comment (ExE-Boss) [reply src/java.base/share/classes/java/lang/Process.java:504]: This method needs to be overridden in `test/lib/jdk/test/lib/process/ProcessTools.java` to call `ProcessTools.ProcessImpl::waitForStreams`.

File: src/java.base/share/classes/java/lang/Process.java (hunk 2 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/Process.java:481]: It will be addressed by https://bugs.openjdk.org/browse/JDK-8336679

File: src/java.base/share/classes/java/lang/Process.java (hunk 1 above)
Comment (naotoj) [reply src/java.base/share/classes/java/lang/Process.java:504]: I don't think current `ProcessTools.startProcess()` even calls `waitFor(long, TimeUnit)` let alone `waitFor(Duration)`. It is polling the process by itself to implement timeout. But I agree it is a good measure to add the override there, for the test to use it in the future.

File: src/java.base/share/classes/java/lang/Process.java (hunk 2 above)
Comment (naotoj) [reply src/java.base/share/classes/java/lang/Process.java:481]: I am planning to add `@implSpec` with a separate issue: [JDK-8336679](https://bugs.openjdk.org/browse/JDK-8336679)

File: src/java.base/share/classes/java/lang/Process.java [hunk 3]
```diff
@@ -456,14 +457,54 @@ public final BufferedWriter outputWriter(Charset charset) {
...
+     * @throws NullPointerException if duration is null
+     * @since 24
+     */
+    public boolean waitFor(Duration duration)
+            throws InterruptedException {
+        Objects.requireNonNull(duration, "duration"); // throw NPE before other conditions
+
+        if (hasExited())
+            return true;
+        if (duration.isZero() || duration.isNegative())
```
Comment (jaikiran) [inline src/java.base/share/classes/java/lang/Process.java:501]: Hello Naoto, I see that there's a `Duration.isPositive()` API. Should we use that here instead?

File: src/java.base/share/classes/java/lang/Process.java [hunk 4]
```diff
@@ -456,14 +457,54 @@ public final BufferedWriter outputWriter(Charset charset) {
...
+     * @throws InterruptedException if the current thread is interrupted
+     *         while waiting.
+     * @throws NullPointerException if duration is null
+     * @since 24
+     */
+    public boolean waitFor(Duration duration)
+            throws InterruptedException {
+        Objects.requireNonNull(duration, "duration"); // throw NPE before other conditions
+
+        if (hasExited())
```
Comment (jaikiran) [inline src/java.base/share/classes/java/lang/Process.java:499]: For clarity, the javadoc on the private internal `hasExited()` method might need an update - it currently states that the method will be invoked from `waitFor(long, TimeUnit)`.

File: test/jdk/java/lang/Process/WaitForDuration.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            Arguments.of(Duration.ofSeconds(Long.MIN_VALUE), false) // nano underflow
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("durations")
+    void testEdgeDurations(Duration d, boolean expected)
+            throws IOException, InterruptedException {
+        assertEquals(expected,
+            new ProcessBuilder("sleep", "3").start().waitFor(d));
```
Comment (jaikiran) [inline test/jdk/java/lang/Process/WaitForDuration.java:57]: I think in its current form, this has a chance of failure (for inputs like 0 or negative duration), if the sleep (for 3 seconds) completes (and thus the process exits) before the `Process.waitFor` implementation has had a chance to execute `hasExited()`.

Also, this test is marked to run on all platforms. I think we might need special handling for `sleep` executable on Windows. In fact, looking at the `initSleepPath` in the `test/jdk/java/lang/ProcessBuilder/Basic.java` test, I suspect we might need something similar in this test even for *nix.

File: src/java.base/share/classes/java/lang/Process.java (hunk 2 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/Process.java:481]: Okay, just a bit strange to add waitFor(Duration) here and then follow-up immediately to change it to move the text to an implSpec, I had hoped it would be done here.

Comment (liach) [reply src/java.base/share/classes/java/lang/Process.java:481]: I believe the choice was made because both waitFor methods need this documentation update; better not touch the (long,TimeUnit) one in this patch.

File: test/jdk/java/lang/Process/WaitForDuration.java (hunk 5 above)
Comment (naotoj) [reply test/jdk/java/lang/Process/WaitForDuration.java]: Made the sleep length variable, and used one hour for zero or negative (should be enough). For the positive, the process completes immediately. Also, I changed the sleep part to pure Java so that it won't rely on the testing platform.

File: src/java.base/share/classes/java/lang/Process.java [hunk 6]
```diff
@@ -456,14 +457,54 @@ public final BufferedWriter outputWriter(Charset charset) {
      * @since 1.8
      */
     public boolean waitFor(long timeout, TimeUnit unit)
-        throws InterruptedException
-    {
-        long remainingNanos = unit.toNanos(timeout); // throw NPE before other conditions
+        throws InterruptedException {
+        Objects.requireNonNull(unit, "unit"); // throw NPE before other conditions
+
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/Process.java:462]: Seems out of scope for this enhancement. Might be better in the PR to update \@implSpec as minor updates.

File: src/java.base/share/classes/java/lang/Process.java [hunk 7]
```diff
@@ -456,14 +457,54 @@ public final BufferedWriter outputWriter(Charset charset) {
...
+    public boolean waitFor(Duration duration)
+            throws InterruptedException {
+        Objects.requireNonNull(duration, "duration"); // throw NPE before other conditions
+
+        if (hasExited())
+            return true;
+        if (!duration.isPositive())
+            return false;
+
+        return waitForNanos(TimeUnit.NANOSECONDS.convert(duration));
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/Process.java:504]: I'd rather see one of these methods call the other, to make it easy keep the behavior in sync.
The Duration can be extracted to nanos and call the existing method.

Comment (naotoj) [reply src/java.base/share/classes/java/lang/Process.java]: Fixed as you suggested.

**Review by liach — APPROVED** [review]

File: src/java.base/share/classes/java/lang/Process.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/Process.java]: Incidentally the same remark has been raised in https://github.com/openjdk/jdk/pull/20220#discussion_r1683352257 and addressed. Good enough for me! :laughing:

File: src/java.base/share/classes/java/lang/Process.java [hunk 8]
```diff
@@ -475,6 +476,35 @@ public boolean waitFor(long timeout, TimeUnit unit)
...
+     * @return {@code true} if the process has exited and {@code false} if
+     *         the waiting duration elapsed before the process has exited.
+     * @throws InterruptedException if the current thread is interrupted
+     *         while waiting.
+     * @throws NullPointerException if duration is null
+     * @since 24
+     */
+    public boolean waitFor(Duration duration) throws InterruptedException {
+        Objects.requireNonNull(duration, "duration");
+        return waitFor(TimeUnit.NANOSECONDS.convert(duration), TimeUnit.NANOSECONDS);
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/Process.java:505]: With this change, the update to ProcessTools is unnecessary and could be backed out.

Comment (liach) [reply src/java.base/share/classes/java/lang/Process.java:505]: The `ProcessTools` override is a delegating override; I think it still makes sense and should be kept.

Comment (naotoj) [reply src/java.base/share/classes/java/lang/Process.java:505]: I think the test tool should not depend on the internal implementation, so I think it should be kept.

File: src/java.base/share/classes/java/lang/Process.java [hunk 9]
```diff
@@ -475,6 +476,35 @@ public boolean waitFor(long timeout, TimeUnit unit)
...
+     *                this method returns immediately.
+     * @return {@code true} if the process has exited and {@code false} if
+     *         the waiting duration elapsed before the process has exited.
+     * @throws InterruptedException if the current thread is interrupted
+     *         while waiting.
+     * @throws NullPointerException if duration is null
+     * @since 24
+     */
+    public boolean waitFor(Duration duration) throws InterruptedException {
+        Objects.requireNonNull(duration, "duration");
```
Comment (pavelrappo) [inline src/java.base/share/classes/java/lang/Process.java:504]: Does this explicit null-check pull its weight here?

Comment (liach) [reply src/java.base/share/classes/java/lang/Process.java:504]: Maybe not, but when this method gets a custom implementation, people don't have to remember to add a null check at front so there's no path that misses the NPE. It isn't wrong, so I don't see a problem with this.

Comment (naotoj) [reply src/java.base/share/classes/java/lang/Process.java:504]: Maybe it's a personal preference, but I think it is a good practice to explicitly check non-null for public APIs. Some recent methods in that class, such as `inputReader(Charset)` do it, so there are precedences.

**Review by jaikiran — APPROVED** [review]
Thank you Naoto, the updated changes look good to me.

File: src/java.base/share/classes/java/lang/Process.java (hunk 8 above)
Comment (RogerRiggs) [reply src/java.base/share/classes/java/lang/Process.java:505]: Its still unnecessary, adds bulk and maintenance overhead that will go unused.

**Review by RogerRiggs — APPROVED** [review]

File: src/java.base/share/classes/java/lang/Process.java (hunk 8 above)
Comment (naotoj) [reply src/java.base/share/classes/java/lang/Process.java:505]: I will leave it as is for now, as it is consistent with other `waitFor` overrides.

**Comment by naotoj** [comment]
Thanks for the reviews!
/integrate
